pyomo/core/plugins/transform/discrete_vars.py

```diff
@@ -25,56 +41,181 @@
             for v, d in model._relaxed_integer_vars[None].values():
                 bounds = v.bounds
                 v.domain = d
                 v.setlb(bounds[0])
                 v.setub(bounds[1])
             model.del_component("_relaxed_integer_vars")
             return
-        # True by default, you can specify False if you want
-        descend = kwds.get(
-            'transform_deactivated_blocks',
-            options.get('transform_deactivated_blocks', True),
-        )
-        active = None if descend else True
 
-        # Relax the model
-        relaxed_vars = {}
-        _base_model_vars = model.component_data_objects(
-            Var, active=active, descend_into=True
-        )
-        for var in _base_model_vars:
+        targets = (model,) if config.targets is None else config.targets
+
+        if config.reverse is None:
+            reverse_dict = {}
+            # Relax the model
+            reverse_token = ReverseTransformationToken(
+                self.__class__, model, targets, reverse_dict
+            )
+        else:
+            # reverse the transformation
+            reverse_token = config.reverse
+            reverse_token.check_token_valid(self.__class__, model, targets)
+            reverse_dict = reverse_token.reverse_dict
+            for v, d in reverse_dict.values():
+                lb, ub = v.bounds
+                v.domain = d
+                v.setlb(lb)
+                v.setub(ub)
+            return
+
+        ### [ESJ 4/29/25]: This can go away when we remove 'undo'
+        model._relaxed_integer_vars = Suffix(direction=Suffix.LOCAL)
+        model._relaxed_integer_vars[None] = reverse_dict
+        ###
+
+        for t in targets:
+            if isinstance(t, Block):
+                blocks = t.values() if t.is_indexed() else (t,)
+                for block in blocks:
+                    self._relax_block(block, config, reverse_dict)
+            elif t.ctype is Var:
+                self._relax_var(t, reverse_dict)
+            else:
+                raise ValueError(
+                    "Target '%s' was not a Block or Var. It was of type "
+                    "'%s' and cannot be transformed." % (t.name, type(t))
+                )
+
+        return reverse_token
+
+    def _relax_block(self, block, config, reverse_dict):
+        self._relax_vars_from_block(block, config, reverse_dict)
+
+        for b in block.component_data_objects(Block, active=None, descend_into=True):
+            if not b.active:
+                if config.transform_deactivated_blocks:
+                    deprecation_warning(
+                        "The `transform_deactivated_blocks` arguments is deprecated. "
+                        "Either specify deactivated Blocks as targets to activate them "
+                        "if transforming them is the desired behavior.",
+                        version='6.9.3.dev0',
+                    )
+                else:
+                    continue
```

Comment:
Some questions:
1. Why descend only over `Block` when you accept `Block` and `Disjunct` as targets?  Should this be `component_data_objects(SubclassOf(Block), ...`?
2. Why test for `active` inside the loop?  Wouldn't it be easier to `active = None if config.transform_deactivated_blocks else True`?  Is the idea to only issue the deprecation warning if model actually had a deactivated block?

Reply:
Yes, all of this is to make sure we only issue the deprecation warning if the model actually has a deactivated block that we're transforming. When we remove the argument, all this complexity can go away and we can just do what you suggest above. I just didn't want to be loud in all the cases where this is fine because (annoyingly) the default value for `transform_deactivated_blocks` is `True`. So without this mess I would just be raising deprecation warnings when the user didn't actually *do* anything--they probably don't know the option exists.